### PR TITLE
Fix Wishart/InverseWishart ambiguities (for Tracker)

### PR DESCRIPTION
**Pull request by nmheim**
Also updates the CI badges in the README and moves and adjoint to `src/zygote.jl` that slipped through #95

### CODE REVIEW

File: src/reversediff.jl [hunk 1]
```diff
@@ -255,15 +255,23 @@ function logpdf(d::MatrixBeta, X::AbstractArray{<:TrackedMatrix{<:Real}})
...
+Distributions.Wishart(df::TrackedReal, S::AbstractPDMat{<:TrackedReal}) = TuringWishart(df, S)
+
+Distributions.InverseWishart(df::TrackedReal, S::Matrix{<:Real}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::AbstractMatrix{<:Real}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::Real, S::AbstractMatrix{<:TrackedReal}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::AbstractMatrix{<:TrackedReal}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::Real, S::TrackedMatrix) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::TrackedMatrix) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::Real, S::AbstractPDMat{<:TrackedReal}) = TuringInverseWishart(df, S)
+Distributions.InverseWishart(df::TrackedReal, S::AbstractPDMat{<:TrackedReal}) = TuringInverseWishart(df, S)
```
Comment (devmotion) [inline src/reversediff.jl:274]: Revert these changes? According to the title the PR is supposed to fix only Tracker issues, and I think it's better to keep these changes for the separate ReverseDiff PR (where I just added some comments).

File: src/tracker.jl [hunk 2]
```diff
@@ -518,10 +518,7 @@ end
 function Distributions.logpdf(d::Wishart, X::AbstractArray{<:TrackedMatrix})
     return logpdf(TuringWishart(d), X)
 end
-Distributions.Wishart(df::TrackedReal, S::Matrix{<:Real}) = TuringWishart(df, S)
-Distributions.Wishart(df::TrackedReal, S::AbstractMatrix{<:Real}) = TuringWishart(df, S)
```
Comment (devmotion) [inline src/tracker.jl:522]: Why are these not needed anymore? It seems they are not covered by the dispatch below.

File: src/tracker.jl [hunk 3]
```diff
@@ -518,10 +518,7 @@ end
...
 Distributions.Wishart(df::Real, S::TrackedMatrix) = TuringWishart(df, S)
-Distributions.Wishart(df::TrackedReal, S::TrackedMatrix) = TuringWishart(df, S)
```
Comment (devmotion) [inline src/tracker.jl:524]: Same here, why can we drop it?

Comment (devmotion) [reply src/tracker.jl:524]: It seems instead of dropping these constructors we should fix the ambiguity by adding explicit constructors for `AbstractPDMat`, similar to the ReverseDiff PR, such as
```julia
Distributions.Wishart(df::TrackedReal, S::AbstractPDMat{<:TrackedReal}) = TuringWishart(df, S) 
```

**Comment by nmheim** [comment]
The deleted constructors caused ambiguities such as:
```julia
Matrixvariate distributions: Error During Test at /home/niklas/.julia/dev/DistributionsAD/test/ad/utils.jl:141                                                                                                       
  Test threw exception                                                                                                                                                                                               
  Expression: ≈(Tracker.data((Tracker.gradient(f, x))[1]), finitediff, rtol=rtol, atol=atol)                                                                                                                         
  MethodError: Wishart(::Tracker.TrackedReal{Float64}, ::ScalMat{Tracker.TrackedReal{Float64}}) is ambiguous. Candidates:                                                                                            
    (::Type{Wishart})(df::T, S::AbstractPDMat{T}) where T<:Real in Distributions at /home/niklas/.julia/packages/Distributions/vskB8/src/matrix/wishart.jl:46                                                        
    (::Type{Wishart})(df::Tracker.TrackedReal, S::AbstractArray{#s17,2} where #s17<:Real) in DistributionsAD at /home/niklas/.julia/dev/DistributionsAD/src/tracker.jl:522                                           
    (::Type{Wishart})(df::Real, S::AbstractPDMat) in Distributions at /home/niklas/.julia/packages/Distributions/vskB8/src/matrix/wishart.jl:62                                                                      
  Possible fix, define                                                                                                                                                                                               
    Wishart(::T<:Tracker.TrackedReal, ::AbstractPDMat{T<:Tracker.TrackedReal})
```

and if I define that I get:
```julia
Matrixvariate distributions: Error During Test at /home/niklas/.julia/dev/DistributionsAD/test/ad/utils.jl:141                                                                                                       
  Test threw exception                                                                                                                                                                                               
  Expression: ≈(Tracker.data((Tracker.gradient(f, x))[1]), finitediff, rtol=rtol, atol=atol)                                                                                                                         
  MethodError: no method matching Float64(::Tracker.TrackedReal{Float64})                                                                                                                                            
  Closest candidates are:                                                                                                                                                                                            
    Float64(::Real, !Matched::RoundingMode) where T<:AbstractFloat at rounding.jl:200
    Float64(::T) where T<:Number at boot.jl:718
    Float64(!Matched::Int8) at float.jl:60
    ...
  Stacktrace:
   [1] convert(::Type{Float64}, ::Tracker.TrackedReal{Float64}) at ./number.jl:7
   [2] setindex!(::Array{Float64,1}, ::Tracker.TrackedReal{Float64}, ::Int64) at ./array.jl:782
   [3] setindex! at ./multidimensional.jl:488 [inlined]
   [4] copyto! at ./multidimensional.jl:905 [inlined]
   [5] copyto! at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/diagonal.jl:73 [inlined]
   [6] copymutable at ./abstractarray.jl:899 [inlined]
   [7] copy at ./abstractarray.jl:849 [inlined]
   [8] copy_oftype(::Diagonal{Float64,TrackedArray{…,Array{Float64,1}}}, ::Type{Float64}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/LinearAlgebra.jl:347
   [9] cholcopy at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/cholesky.jl:139 [inlined]
   [10] #cholesky#158 at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/diagonal.jl:663 [inlined]
   [11] cholesky at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/diagonal.jl:663 [inlined] (repeats 2 times)
   [12] cholesky(::ScalMat{Tracker.TrackedReal{Float64}}) at /home/niklas/.julia/packages/PDMats/8bndI/src/scalmat.jl:21
   [13] TuringWishart at /home/niklas/.julia/dev/DistributionsAD/src/matrixvariate.jl:25 [inlined]
   [14] Wishart at /home/niklas/.julia/dev/DistributionsAD/src/tracker.jl:522 [inlined]
   [15] MatrixBeta(::Int64, ::Tracker.TrackedReal{Float64}, ::Float64) at /home/niklas/.julia/packages/Distributions/vskB8/src/matrix/matrixbeta.jl:43
   [16] (::var"#82#161"{Int64})(::Tracker.TrackedReal{Float64}, ::Float64) at /home/niklas/.julia/dev/DistributionsAD/test/ad/distributions.jl:320
   [17] (::var"#17#21"{var"#82#161"{Int64},var"#83#162"{var"#to_beta_mat#102"}})(::Array{Array{Float64,2},1}, ::Tracker.TrackedReal{Float64}, ::Vararg{Any,N} where N) at /home/niklas/.julia/dev/DistributionsAD/tes
t/ad/utils.jl:97
   [18] (::var"#19#23"{var"#17#21"{var"#82#161"{Int64},var"#83#162"{var"#to_beta_mat#102"}},Array{Array{Float64,2},1},Tuple{Float64,Float64},Array{Int64,1}})(::TrackedArray{…,Array{Float64,1}}) at /home/niklas/.ju
lia/dev/DistributionsAD/test/ad/utils.jl:116
   [19] gradient_(::Function, ::Array{Float64,1}) at /home/niklas/.julia/packages/Tracker/32Vek/src/back.jl:90
   [20] #gradient#24 at /home/niklas/.julia/packages/Tracker/32Vek/src/back.jl:164 [inlined]
   [21] gradient at /home/niklas/.julia/packages/Tracker/32Vek/src/back.jl:164 [inlined]
   [22] #test_ad#25(::Float64, ::Float64, ::typeof(test_ad), ::Function, ::Array{Float64,1}, ::Tuple{}) at /home/niklas/.julia/dev/DistributionsAD/test/ad/utils.jl:141
   [23] test_ad(::Function, ::Array{Float64,1}, ::Tuple{}) at /home/niklas/.julia/dev/DistributionsAD/test/ad/utils.jl:135
...
```

**Comment by devmotion** [comment]
I see, but we really want to dispatch to `TuringWishart` (it's type piracy but it's a workaround for the time being...). I assume in addition to the additional method you might have to add something similar to the ReverseDiff PR (I didn't really understand why it's needed but maybe that's the reason...) and replace https://github.com/nmheim/DistributionsAD.jl/blob/5547c2a529b403a03bf4b746592f55d5e3712586/src/matrixvariate.jl#L18-L27 with
```julia
getchol(p::AbstractMatrix) = cholesky(p)
getchol(p::PDMat) = p.chol
getchol(p::PDiagMat) = Cholesky(Diagonal(sqrt.(p.diag)), 'U', 0)
getchol(p::ScalMat) = Cholesky(Diagonal(fill(sqrt(p.value), p.dim)), 'U', 0) 

function TuringWishart(df::Real, S::AbstractMatrix)
    p = size(S, 1)
    df > p - 1 || error("dpf should be greater than dim - 1.")
    C = getchol(S)
    return TuringWishart(df, C)
end
```
Ideally one would just use `cholesky` instead of `getchol` but there seems to be some problem with the PDMats and Tracker.

File: src/tracker.jl (hunk 3 above)
Comment (nmheim) [reply src/tracker.jl:524]: `TrackedReal` is a subtype of `Real`, so that constructor should not be necessary?

Comment (devmotion) [reply src/tracker.jl:524]: The problem is the ambiguity error that arises from having both `Wishart(::Real, ::AbstractMatrix{<:Real})` defined above and `Wishart(::T, ::AbstractPDMat{<:T}) where {T<:Real}` in https://github.com/JuliaStats/Distributions.jl/blob/master/src/matrix/wishart.jl#L45 since `AbstractPDMat <: AbstractMatrix`.

**Comment by nmheim** [comment]
Ok that does it! :)

Just so I understand: Why do the tests not fail without the additional constructors though? Is `Wishart` differentiable after all? Or is it taken care of  https://github.com/nmheim/DistributionsAD.jl/blob/fd68d10feb8eade08ff0df8f2ff73bc58bbcb6f0/src/tracker.jl#L515-L520 and its just a performance issue?

**Comment by devmotion** [comment]
> Or is it taken care of nmheim/DistributionsAD.jl@fd68d10/src/tracker.jl#L515-L520 and its just a performance issue?

I assume that's the case. Maybe @mohamed82008 has some additional insights?

**Comment by nmheim** [comment]
Yes, sorry, the request for review was a misclick. I was just checking something similar. The tests pass both with and without your suggested changes. This is the same puzzle as above I guess... I added your suggested changes for `InverseWishart`, but if they are not needed I can revert them again...

**Comment by devmotion** [comment]
Let's keep them for now (for consistency and since the other PR included similar changes as well). Maybe we just have to extend our tests.

Thanks for the fixes!
